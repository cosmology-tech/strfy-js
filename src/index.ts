export * from './js';
export * from './json';