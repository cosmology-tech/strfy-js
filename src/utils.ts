import { minimatch } from 'minimatch';

export function camelCaseTransform(key: string): string {
  return key.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
}

export function isSimpleKey(key: string): boolean {
  return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key);
}

export function dirname(path: string): string {
  return path.replace(/\/[^\/]*$/, ''); // Removes last segment after the last '/'
}

export function escapeStringForSingleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only single quotes
  str = str.replace(/'/g, "\\'");
  return str;
}

export function escapeStringForDoubleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only double quotes
  str = str.replace(/"/g, '\\"');
  return str;
}


export function escapeStringForBacktickQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape backticks
  str = str.replace(/`/g, '\\`');
  return str;
}


const globPattern = /\*+([^+@!?\*\[\(]*)/;

interface ShouldIncludeOptions {
  include: string[];
  exclude: string[];
}

export function expandPaths(paths: string[]): string[] {
  const pathSet = new Set<string>();

  paths.forEach(path => {
    let currentPath = path;
    // Ensure the path is added to the set
    pathSet.add(currentPath);

    // Generate and add all parent directories to the set
    while (currentPath !== '') {
      const lastSlashIndex = currentPath.lastIndexOf('/');
      if (lastSlashIndex === -1) break; // No more slashes, stop processing

      currentPath = currentPath.substring(0, lastSlashIndex) || '';
      if (currentPath !== '')
        pathSet.add(currentPath);
    }
  });

  return sortPaths(Array.from(pathSet));
}

function sortPaths(paths) {
  // First sort alphabetically
  paths.sort((a, b) => a.localeCompare(b));

  // Then sort by depth (number of '/' in the path)
  paths.sort((a, b) => {
    const depthA = (a.match(/\//g) || []).length;
    const depthB = (b.match(/\//g) || []).length;
    return depthB - depthA;  // Sort from deepest to shallowest
  });

  return paths;
}


function hasAncestor(currentPath, includePaths) {
  return !!findNearestAncestor(currentPath, includePaths);
}

function findNearestAncestor(currentPath, includePaths) {
  let pathToCheck = currentPath;
  while (pathToCheck !== '/' && pathToCheck !== '') {
    if (includePaths.some(includePath => minimatch(pathToCheck + '/', includePath))) {
      if (pathToCheck !== currentPath)
        return pathToCheck; // Return the matching parent directory
    }
    pathToCheck = dirname(pathToCheck); // Move to the next parent directory
  }
  return null; // Return null if no matching parent is found
}

export const shouldInclude = (type: string, options: ShouldIncludeOptions): boolean => {
  // Determine if 'include' and 'exclude' are effectively set
  const includesEffectivelySet = options.include && options.include.length > 0;
  const excludesEffectivelySet = options.exclude && options.exclude.length > 0;

  // Function to check if any patterns in the array match the type
  const matchesPattern = (patterns: string[], type: string): boolean =>
    patterns.some(pattern => globPattern.test(pattern) ? minimatch(type, pattern) : type === pattern);

  const isIncluded = includesEffectivelySet ? matchesPattern(expandPaths(options.include), type) : true;
  const isIncludedByAncestor = !!findNearestAncestor(type, options.include);
  const isExcluded = excludesEffectivelySet ? matchesPattern(options.exclude, type) : false;

  const toInclude = isIncluded;

  // Apply the logic based on whether includes or excludes are effectively set
  if (includesEffectivelySet && excludesEffectivelySet) {
    if (isIncludedByAncestor && !isExcluded) {
      return true;
    }
    return toInclude && !isExcluded;
  } else if (includesEffectivelySet) {
    if (isIncludedByAncestor) {
      return true;
    }
    return toInclude;
  } else if (excludesEffectivelySet) {
    return !isExcluded;
  }

  // Default behavior if neither is effectively set
  return true;
}