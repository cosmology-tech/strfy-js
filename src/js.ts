import { camelCaseTransform, escapeStringForBacktickQuotes, escapeStringForDoubleQuotes, escapeStringForSingleQuotes, isSimpleKey } from "./utils";

export interface JSStringifyOptions {
  space?: number;
  replacer?: (key: string, value: any) => any | null;
  propertyReplacer?: (currentKey: string, currentPath: string, propertyRenameMap: { [pattern: string]: string }) => string
  quotes?: 'single' | 'double' | 'backtick';
  inlineArrayLimit?: number;
  camelCase?: boolean;
  camelCaseFn?: (str: string) => string;
  propertyRenameMap?: { [key: string]: string };
  json?: boolean;
}


export function chooseQuotes(str: string, preferred: 'single' | 'double' | 'backtick'): string {
  switch (preferred) {
    case 'single':
      return `'${escapeStringForSingleQuotes(str)}'`;
    case 'double':
      return `"${escapeStringForDoubleQuotes(str)}"`;
    case 'backtick':
      return `\`${escapeStringForBacktickQuotes(str)}\``;
    default:
      throw new Error("Invalid quote type specified.");
  }
}

function matchesPattern(path, pattern) {
  const pathSegments = path.split('/');
  const patternSegments = pattern.split('/');
  if (pathSegments.length !== patternSegments.length) {
    return false;
  }
  for (let i = 0; i < pathSegments.length; i++) {
    if (patternSegments[i] !== '*' && patternSegments[i] !== pathSegments[i]) {
      return false;
    }
  }
  return true;
}

function defaultPropertyReplacer(
  currentKey: string,
  currentPath: string,
  propertyRenameMap: { [pattern: string]: string }
): string {
  let finalKey = currentKey;
  Object.keys(propertyRenameMap).forEach(pattern => {
    if (matchesPattern(currentPath, pattern)) {
      finalKey = propertyRenameMap[pattern];
    }
  });
  return finalKey;
}


export function jsStringify(obj: any, options?: JSStringifyOptions): string {
  const {
    space = 0,
    replacer = null,
    propertyReplacer = null,
    quotes = 'single',
    inlineArrayLimit = undefined,
    camelCase = false,
    camelCaseFn = camelCaseTransform,
    propertyRenameMap = {},
    json = false  // JSON compliance mode
  } = options || {};

  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);

  let indentLevel: number = 0;

  const serialize = (obj: any, currentPath = "", isArray: boolean = false): string => {
    if (replacer instanceof Function) {
      obj = replacer.call(null, '', obj); // Call replacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map((item, index) => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(item, `${currentPath}/${index}`, true);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;
      const props = Object.keys(obj).map(key => {
        const fullPath = `${currentPath}/${key}`;
        const value = replacer instanceof Function ? replacer.call(obj, key, obj[key]) : obj[key];


        let replacedKey: string;
        switch (true) {
          case propertyReplacer instanceof Function:
            replacedKey = propertyReplacer.call(obj, key, fullPath, propertyRenameMap);
            break;
          default:
            replacedKey = defaultPropertyReplacer(key, fullPath, propertyRenameMap);
        }

        const finalKey = camelCase ? camelCaseFn(replacedKey) : replacedKey;
        const keyPart = json ? `"${finalKey}"` : isSimpleKey(finalKey) ? finalKey : `"${finalKey}"`;
        const valuePart = serialize(value, fullPath);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return json ? `"${escapeStringForDoubleQuotes(obj)}"` : chooseQuotes(obj, quotes);
    } else {
      return String(obj);
    }
  };

  return serialize(obj);
}
