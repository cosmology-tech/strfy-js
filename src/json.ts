import { jsStringify, JSStringifyOptions } from "./js";

export interface JSONStringifyOptions extends Omit<JSStringifyOptions, 'json'> { }

export function jsonStringify(obj: any, options?: JSONStringifyOptions): string {
  // Set json option to true and spread the rest of the options
  return jsStringify(obj, { ...options, json: true });
}
