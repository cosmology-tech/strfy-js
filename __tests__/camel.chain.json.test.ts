import { jsonStringify } from "../src";

import chain from '../__fixtures__/chain.json';

it('chain', () => {
  expect(jsonStringify(chain, { camelCase: true, space: 2 })).toMatchSnapshot();
});