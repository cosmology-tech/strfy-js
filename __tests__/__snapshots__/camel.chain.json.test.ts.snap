// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chain 1`] = `
"{
  "$schema": "../chain.schema.json",
  "chainName": "comdex",
  "status": "live",
  "networkType": "mainnet",
  "website": "https://comdex.one/",
  "prettyName": "Comdex",
  "chainId": "comdex-1",
  "bech32Prefix": "comdex",
  "slip44": 118,
  "fees": {
    "feeTokens": [
      {
        "denom": "ucmdx",
        "fixedMinGasPrice": 0,
        "lowGasPrice": 0,
        "averageGasPrice": 0.025,
        "highGasPrice": 0.04
      }
    ]
  },
  "staking": {
    "stakingTokens": [
      {
        "denom": "ucmdx"
      }
    ]
  },
  "codebase": {
    "gitRepo": "https://github.com/comdex-official/comdex",
    "recommendedVersion": "v13.4.0",
    "compatibleVersions": [
      "v13.4.0"
    ],
    "binaries": {
      "linux/amd64": "https://github.com/comdex-official/comdex/releases/download/v13.4.0/comdex-linux-amd64.tar.gz"
    },
    "cosmosSdkVersion": "v0.47.5",
    "consensus": {
      "type": "cometbft",
      "version": "0.37.2"
    },
    "cosmwasmVersion": "v0.41.0",
    "cosmwasmEnabled": true,
    "genesis": {
      "genesisUrl": "https://comdex-mainnet-genesis.s3.ap-southeast-1.amazonaws.com/genesis.json"
    },
    "versions": [
      {
        "name": "v9.0.0",
        "recommendedVersion": "v9.0.0",
        "compatibleVersions": [
          "v9.0.0"
        ],
        "cosmwasmEnabled": true,
        "nextVersionName": "v10.0.0"
      },
      {
        "name": "v10.0.0",
        "recommendedVersion": "v10.0.0",
        "compatibleVersions": [
          "v10.0.0"
        ],
        "cosmwasmEnabled": true,
        "nextVersionName": "v11.5.0"
      },
      {
        "name": "v11.5.0",
        "height": 8184000,
        "proposal": 154,
        "recommendedVersion": "v11.5.2",
        "compatibleVersions": [
          "v11.5.2"
        ],
        "cosmwasmEnabled": true,
        "nextVersionName": "v13.3.0"
      },
      {
        "name": "v13.3.0",
        "proposal": 216,
        "height": 10981900,
        "recommendedVersion": "v13.4.0",
        "compatibleVersions": [
          "v13.4.0"
        ],
        "binaries": {
          "linux/amd64": "https://github.com/comdex-official/comdex/releases/download/v13.4.0/comdex-linux-amd64.tar.gz"
        },
        "cosmosSdkVersion": "v0.47.5",
        "consensus": {
          "type": "cometbft",
          "version": "0.37.2"
        },
        "cosmwasmVersion": "v0.41.0",
        "cosmwasmEnabled": true,
        "nextVersionName": ""
      }
    ]
  },
  "daemonName": "comdex",
  "nodeHome": "$HOME/.comdex",
  "keyAlgos": [
    "secp256k1"
  ],
  "logoURIs": {
    "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png",
    "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg"
  },
  "peers": {
    "seeds": [
      {
        "id": "7ca14a1d156299999eba9c394ca060368022d52f",
        "address": "54.194.178.110:26656"
      },
      {
        "id": "ade4d8bc8cbe014af6ebdf3cb7b1e9ad36f412c0",
        "address": "seeds.polkachu.com:13156",
        "provider": "Polkachu"
      },
      {
        "id": "c30dacf15e97a30b78792c7fa817fd2ef529b820",
        "address": "comdex.seed.stavr.tech:2026",
        "provider": "üî•STAVRüî•"
      },
      {
        "id": "20e1000e88125698264454a884812746c2eb4807",
        "address": "seeds.lavenderfive.com:13156",
        "provider": "Lavender.Five Nodes üêù"
      },
      {
        "id": "ebc272824924ea1a27ea3183dd0b9ba713494f83",
        "address": "comdex-mainnet-seed.autostake.com:26776",
        "provider": "AutoStake üõ°Ô∏è Slash Protected"
      },
      {
        "id": "6596d143fd128b2394b27cd7846bda099ca5a193",
        "address": "seeds.goldenratiostaking.net:1621",
        "provider": "Golden Ratio Staking"
      },
      {
        "id": "8542cd7e6bf9d260fef543bc49e59be5a3fa9074",
        "address": "seed.publicnode.com:26656",
        "provider": "Allnodes ‚ö°Ô∏è Nodes & Staking"
      },
      {
        "id": "c28827cb96c14c905b127b92065a3fb4cd77d7f6",
        "address": "seeds.whispernode.com:13156",
        "provider": "WhisperNode ü§ê"
      },
      {
        "id": "88ba33fbdf0279efaf27cff629f3cf72814d4069",
        "address": "seed-comdex-01.stakeflow.io:10007",
        "provider": "Stakeflow"
      },
      {
        "id": "df949a46ae6529ae1e09b034b49716468d5cc7e9",
        "address": "seeds.stakerhouse.com:10456",
        "provider": "StakerHouse"
      }
    ],
    "persistentPeers": [
      {
        "id": "d478882a80674fa10a32da63cc20cae13e3a2a57",
        "address": "43.204.0.243:26656"
      },
      {
        "id": "d8b74791ee56f1b345d822f62bd9bc969668d8df",
        "address": "194.163.128.55:36656"
      },
      {
        "id": "ebc272824924ea1a27ea3183dd0b9ba713494f83",
        "address": "comdex-mainnet-peer.autostake.com:26776",
        "provider": "AutoStake üõ°Ô∏è Slash Protecteds"
      },
      {
        "id": "88ba33fbdf0279efaf27cff629f3cf72814d4069",
        "address": "peer-comdex-01.stakeflow.io:10007",
        "provider": "Stakeflow"
      }
    ]
  },
  "apis": {
    "rpc": [
      {
        "address": "https://rpc.comdex.one",
        "provider": "comdex"
      },
      {
        "address": "https://comdex-rpc.polkachu.com",
        "provider": "Polkachu"
      },
      {
        "address": "https://comdex.rpc.m.stavr.tech",
        "provider": "üî•STAVRüî•"
      },
      {
        "address": "https://rpc.comdex.chaintools.tech/",
        "provider": "ChainTools"
      },
      {
        "address": "https://comdex-rpc.lavenderfive.com/",
        "provider": "Lavender.Five Nodes üêù"
      },
      {
        "address": "https://rpc-comdex.cosmos-spaces.cloud",
        "provider": "Cosmos Spaces"
      },
      {
        "address": "https://rpc-comdex.whispernode.com:443",
        "provider": "WhisperNode ü§ê"
      },
      {
        "address": "https://comdex-mainnet-rpc.autostake.com:443",
        "provider": "AutoStake üõ°Ô∏è Slash Protected"
      },
      {
        "address": "https://comdex-rpc.w3coins.io",
        "provider": "w3coins"
      },
      {
        "address": "https://comdex-rpc.stakerhouse.com",
        "provider": "StakerHouse"
      },
      {
        "address": "https://comdex-rpc.publicnode.com:443",
        "provider": "Allnodes ‚ö°Ô∏è Nodes & Staking"
      },
      {
        "address": "https://rpc-comdex-01.stakeflow.io",
        "provider": "Stakeflow"
      },
      {
        "address": "https://comdex-rpc.validatornode.com",
        "provider": "ValidatorNode"
      }
    ],
    "rest": [
      {
        "address": "https://rest.comdex.one",
        "provider": "comdex"
      },
      {
        "address": "https://api.comdex.chaintools.tech/",
        "provider": "ChainTools"
      },
      {
        "address": "https://comdex-api.polkachu.com",
        "provider": "Polkachu"
      },
      {
        "address": "https://comdex.api.m.stavr.tech",
        "provider": "üî•STAVRüî•"
      },
      {
        "address": "https://comdex-api.lavenderfive.com/",
        "provider": "Lavender.Five Nodes üêù"
      },
      {
        "address": "https://api-comdex.cosmos-spaces.cloud",
        "provider": "Cosmos Spaces"
      },
      {
        "address": "https://comdex-mainnet-lcd.autostake.com:443",
        "provider": "AutoStake üõ°Ô∏è Slash Protected"
      },
      {
        "address": "https://lcd-comdex.whispernode.com:443",
        "provider": "WhisperNode ü§ê"
      },
      {
        "address": "https://comdex-api.w3coins.io",
        "provider": "w3coins"
      },
      {
        "address": "https://comdex-rest.stakerhouse.com",
        "provider": "StakerHouse"
      },
      {
        "address": "https://comdex-rest.publicnode.com",
        "provider": "Allnodes ‚ö°Ô∏è Nodes & Staking"
      },
      {
        "address": "https://api-comdex-01.stakeflow.io",
        "provider": "Stakeflow"
      },
      {
        "address": "https://comdex-api.validatornode.com",
        "provider": "ValidatorNode"
      }
    ],
    "grpc": [
      {
        "address": "grpc-comdex-ia.cosmosia.notional.ventures:443",
        "provider": "Notional"
      },
      {
        "address": "comdex.grpcui.chaintools.host:443",
        "provider": "ChainTools"
      },
      {
        "address": "comdex-grpc.polkachu.com:13190",
        "provider": "Polkachu"
      },
      {
        "address": "comdex.grpc.m.stavr.tech:104",
        "provider": "üî•STAVRüî•"
      },
      {
        "address": "grpc-comdex.cosmos-spaces.cloud:2300",
        "provider": "Cosmos Spaces"
      },
      {
        "address": "comdex-grpc.lavenderfive.com:443",
        "provider": "Lavender.Five Nodes üêù"
      },
      {
        "address": "comdex-mainnet-lcd.autostake.com:443",
        "provider": "AutoStake üõ°Ô∏è Slash Protected"
      },
      {
        "address": "comdex-grpc.w3coins.io:13190",
        "provider": "w3coins"
      },
      {
        "address": "comdex-grpc.publicnode.com:443",
        "provider": "Allnodes ‚ö°Ô∏è Nodes & Staking"
      },
      {
        "address": "grpc-comdex-01.stakeflow.io:10002",
        "provider": "Stakeflow"
      },
      {
        "address": "comdex-grpc.stakerhouse.com:443",
        "provider": "StakerHouse"
      }
    ]
  },
  "explorers": [
    {
      "kind": "ezstaking",
      "url": "https://ezstaking.app/comdex",
      "txPage": "https://ezstaking.app/comdex/txs/\${txHash}",
      "accountPage": "https://ezstaking.app/comdex/account/\${accountAddress}"
    },
    {
      "kind": "mintscan",
      "url": "https://www.mintscan.io/comdex",
      "txPage": "https://www.mintscan.io/comdex/transactions/\${txHash}",
      "accountPage": "https://www.mintscan.io/comdex/accounts/\${accountAddress}"
    },
    {
      "kind": "üî•STAVRüî•",
      "url": "https://explorer.stavr.tech/Comdex-Mainnet",
      "txPage": "https://explorer.stavr.tech/Comdex-Mainnet/txs/\${txHash}",
      "accountPage": "https://explorer.stavr.tech/Comdex-Mainnet/accounts/\${accountAddress}"
    },
    {
      "kind": "aneka",
      "url": "https://comdex.aneka.io/",
      "txPage": "https://comdex.aneka.io/txs/\${txHash}"
    },
    {
      "kind": "ping.pub",
      "url": "https://ping.pub/comdex",
      "txPage": "https://ping.pub/comdex/tx/\${txHash}"
    },
    {
      "kind": "atomscan",
      "url": "https://atomscan.com/comdex",
      "txPage": "https://atomscan.com/comdex/transactions/\${txHash}",
      "accountPage": "https://atomscan.com/comdex/accounts/\${accountAddress}"
    },
    {
      "kind": "Stakeflow",
      "url": "https://stakeflow.io/comdex",
      "accountPage": "https://stakeflow.io/comdex/accounts/\${accountAddress}"
    },
    {
      "kind": "ValidatorNode",
      "url": "https://explorer.validatornode.com/comdex",
      "txPage": "https://explorer.validatornode.com/comdex/tx/\${txHash}"
    }
  ],
  "images": [
    {
      "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png",
      "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg"
    }
  ]
}"
`;
